HDDS-6909. [Multi-Tenant] Use RangerClient for Ranger operations

smengcl opened the pull request:
## What changes were proposed in this pull request?

- [x] Use `RangerClient` for Ranger read and write operations.
- [x] Add `ozone admin om updateranger` command to allow Ozone admins to manually trigger BG Sync to Ranger.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6909

## How was this patch tested?

- [x] All existing tests should pass.
- [x] `TestRangerBGSyncService` passed when connected to a remote Ranger Admin Server.

smengcl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java:
```diff
@@ -267,6 +267,11 @@ public static boolean isReadOnly(
+    case RangerBGSync:
+      // RangerBGSync is a read operation in the sense that it doesn't directly
+      // write to OM DB. And therefore it doesn't need a OMClientRequest.
+      // Although indirectly the Ranger sync service task could invoke write
```
Right. It could indirectly trigger another write operation. But for the implementation `RangerBGSync` should be listed under the `readonly` list.

kerneltime commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/multitenant/RangerUserRequest.java:
```diff
@@ -0,0 +1,275 @@
+  }
+
+  private void setupRangerIgnoreServerCertificate() {
+    // Create a trust manager that does not validate certificate chains
```
Is there a longer term plan for using customer provided trust chain?

smengcl replied:
Oh. This part is only in the test code and not used in production, and was copied from our earlier homemade Ranger client. This could be useful during dev, though the proper way would be to set a correct truststore.

kerneltime commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/multitenant/RangerUserRequest.java:
```diff
@@ -0,0 +1,275 @@
+
+    urlConnection.setRequestMethod(method);
+    // Timeout in ms
+    urlConnection.setConnectTimeout(5000);
```
ToDo: Make timeout values configurable

smengcl replied:
Yes. The actual production timeout is tunable: https://github.com/apache/ozone/blob/fc3a23af25d61146139ad434b2323bbeedc89828/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/multitenant/RangerRestMultiTenantAccessController.java#L440

`RangerUserRequest` is test-only code so I didn't bother. I could read the some config as well though.

smengcl replied:
Added in 72a4d60

kerneltime commented:
Most of my comments are minor nits or future work, the CI is failing in findbugs
```
M D SF: Switch statement found in org.apache.hadoop.ozone.om.multitenant.RangerClientMultiTenantAccessController.decodeRSEStatusCodes(RangerServiceException) where default case is missing  At RangerClientMultiTenantAccessController.java:[lines 152-157]
``` 
Once it passes we can merge this change in.

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/multitenant/RangerClientMultiTenantAccessController.java:
```diff
@@ -56,138 +60,276 @@ public class RangerClientMultiTenantAccessController implements
+    case HTTP_STATUS_CODE_BAD_REQUEST:
+      LOG.error("Request failure. If this is an assign-user operation, "
+          + "check if the user name exists in Ranger.");
+    default:
```
This is a helper function to translate the return code during error because I find the non-200 status codes often unintuitive from Ranger. (And Ranger sometimes even masks the real reason behind a failure in some operations / cases.)

smengcl commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -3081,6 +3083,53 @@ public ServiceInfoEx getServiceInfo() throws IOException {
+      LOG.info("User '{}' manually triggered Multi-Tenancy Ranger Sync "
+          + "in a new thread, tid={}", ugi, t.getId());
+      return true;
+    } else {
```
Yup. Though I personally like to keep this for symmetry. There is not too much code in the `else` part. If it grows larger I would unindent.

smengcl commented:
> Most of my comments are minor nits or future work, the CI is failing in findbugs
> 
> ```
> M D SF: Switch statement found in org.apache.hadoop.ozone.om.multitenant.RangerClientMultiTenantAccessController.decodeRSEStatusCodes(RangerServiceException) where default case is missing  At RangerClientMultiTenantAccessController.java:[lines 152-157]
> ```
> 
> Once it passes we can merge this change in.

Thanks @kerneltime for the review. I have addressed the comments so far.
